src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,100 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    static final int ARRAY_MIN_SIZE = 10;
+
+    private T[] array = (T[]) new Object[ARRAY_MIN_SIZE];
+    private int size;
+
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            array = (T[]) grow();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            if (size >= array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index, array, index + 1, size - index);
+            array[index] = value;
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("This index is missing");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (isIndexValid(index)) {
+            T removeElement = array[index];
+            System.arraycopy(array, index + 1, array, index, size - index - 1);
+            size--;
+            return removeElement;
+        }
+        throw new ArrayListIndexOutOfBoundsException("This index is missing");
```

Comment:
you will never reach this throw, `isIndexValid()` will never return false, so we can remove this `if`, or maybe even make it void and name it like checkIndex()